### ISSUE TITLE
Custom OAuth should use exactly same redirect_uri for all requests to OAuth provider

### ISSUE THREAD

**Issue by ugoertz** [RocketChat/Rocket.Chat.ReactNative#3045]
Logging into a self-hosted RocketChat via a custom OAuth provider with the mobile app does not work for me.

As far as I understand, it is a strict requirement of the OAuth protocol that the exact same redirect_uri is used for the whole process of authenticating a user, and also query parameters such as `?close` have to match. This is not satisfied with the custom OAuth provider settings, because in the call to the authorize endpoint the `?close` it is not used, but in the call to the token endpoint it is used. (Tested with 4.16.0.23722 on Android.)

It seems that some OAuth provider libraries ignore this issue, but in my case (oauthlib via django-oauth-toolkit) the library is very picky. As a result, the OAuth provider raises an error (mismatch of redirect uri), and users cannot login in this way.

Logs on RocketChat server:
```
W20210408-07:17:42.824(0) (oauth_server.js:403) Error in OAuth Server: Failed to complete OAuth handshake with ... at https://my.oauth.provider/o/token/. failed [400] {"error": "invalid_request", "error_description": "Mismatching redirect URI.", "state": "eyJsb2dpblN0eW..."}
```

Logging in from browsers and the desktop app (tested on Linux) via OAuth works. Hacking the OAuth provider code and manually adding the `?close` in the first incoming request before it is passed to oauthlib also fixes the problem.

I think the correct fix is adding the `?close` in this line:
https://github.com/RocketChat/Rocket.Chat.ReactNative/blob/85f79e7f09f52b3d3cd611bd1ccb1a61ffdee061/app/containers/LoginServices.js#L179
Note that for all other OAuth provides that are defined in the lines above in the same file, the `?close` is present.

Two more remarks.

- The requests from the browser have the characters `:/?` in the redirect_uri urlencoded (`redirect_uri=https%3A%2F%2Fchat...`) while the app sends a non-encoded `redirect_uri=https://chat...`. Both ways work fine (and as far as I understand these characters do not have to be encoded). I did not check what happens when the redirect_uri itself contains a `&` - that not being encoded would cause a problem. In any case, it might be desirable to make this behavior consistent across browser and app.
- In the [RocketChat OAuth documentation](https://docs.rocket.chat/guides/administrator-guides/authentication/oauth) the `?close` parameter is mentioned for some providers, but not for all. If I understand correctly, it is OK with the OAuth specification to add a parameter like this to the preregistered callback URL (but then the parameter has to be present always). Nevertheless, it might be desirable to make this more consistent, e.g. for LinkedIn the app uses `?close`, but in the docs it is not specified.

Thanks for the great work on RocketChat! Thanks for looking into this and please let me know if there is any further info I could provide.

**Comment by diegolmello** [comment]
@ugoertz Can you test on 4.15.0? It's still live on https://play.google.com/store/apps/details?id=chat.rocket.android

**Comment by ugoertz** [comment]
Thanks for the quick reply. I see exactly the same problematic behavior with 4.15.0.23396.

**Comment by diegolmello** [comment]
Ok. It's easy to misconfigure custom oauth and we have a couple of issues about it that may help you https://github.com/RocketChat/Rocket.Chat.ReactNative/issues?q=is%3Aissue+custom+oauth
Please follow them and let us know.

**Comment by ugoertz** [comment]
I do not think any of those issues is relevant, in fact the `redirect_uri` is never mentioned.

I found out that with Login Style set to "Popup" or "Redirect", the redirect_uri used is the same for all requests (never using the `?close` parameter). In particular with those settings the above problem disappears. (Is this in the documentation anywhere?)

With Login Style set to "Default", logging in works with the browser and the desktop app, but not with the mobile client, and I still fail to see any reason why the mobile client behaves differently from the others with this setting.

For my remark about encoding the URI, I now saw that this exact issue was fixed in the browser environment in https://github.com/RocketChat/Rocket.Chat/pull/15957/files - might make sense to do this for the mobile app, too.
